### ISSUE TITLE
Getting different results with DirectML vs CPU or CUDA for Tensorflow Object Detection model

### ISSUE THREAD

**Issue by alex-cyberhaven** [microsoft/onnxruntime#19352]
### Describe the issue

I'm using a fine tuned model based on this SSD architecture https://github.com/tensorflow/models/blob/master/research/object_detection/configs/tf2/ssd_resnet50_v1_fpn_640x640_coco17_tpu-8.config
This was converted to ONNX using tf2onnx.

Inference using the ONNX Runtime built with the CUDA provider gives correct results. When I switch to DML, I get correct results
only when selecting a CPU based D3D device. With any other GPU (tested with integrated Intel, discrete Nvidia, discrete AMD), one of the model output tensors  `detection_scores` is wrong (shows only -1 values). This tensor is computed during the post-processing step of the model https://github.com/tensorflow/models/blob/master/research/object_detection/meta_architectures/ssd_meta_arch.py#L657

So I suspect something in that final processing is not computed properly or maybe even skipped on DML. My first thought was the NonMaxSuppression operator which I saw was used there and it's not implemented in DML but runs on CPU. Could it cause these issues, perhaps due to the partitioning and data being copied?

I tried to troubleshoot in various ways
- Ran with onnxruntime with CUDA: correct output
- Ran with onnxruntime with DML on CPU device: correct output
- Ran with onnxruntime with DML on a GPU: incorrect output
- Disabled optimizations with ORT_DISABLE_ALL: incorrect output
- Compared all the model output tensors between GPU and CPU, only `detection_scores` seems to be garbage
- Made sure to set enable_memory_pattern to false and execution mode to sequential
- With verbose logging and older onnxruntime I found this `QDQ quantization is not supported while using the DML Execution Provider. So disabling it for this session since it uses the DML Execution Provider.` But the model is not quantized and I don't see any related operators in the model.
- Tried using the latest DirectML.dll and the nightly onnxruntime builds
- Tried converting the tensorflow model with different opsets - 15, 16, 17, 18 - with the same result
- Looked for similar Github issues
  - https://github.com/microsoft/onnxruntime/issues/18652 Incorrect inference (DML & Intel Iris Xe GPU)  
  - https://github.com/microsoft/onnxruntime/issues/9582 Divergent results between CPU and DirectML  

Any ideas how to troubleshoot this further to narrow it down? I'm not even sure this is an issue with onnxruntime or DirectML.

### To reproduce

Fine tuned Tensorflow Object Detection model https://github.com/tensorflow/models/blob/master/research/object_detection/configs/tf2/ssd_resnet50_v1_fpn_640x640_coco17_tpu-8.config

Code to fine tune in this tutorial https://github.com/tensorflow/models/blob/master/research/object_detection/colab_tutorials/eager_few_shot_od_training_tf2_colab.ipynb

Convert to ONNX with default options
```
python -m tf2onnx.convert --saved-model .\saved --output model.onnx
```

Tried using these python packages:
- onnx 1.14.0 and tf2onnx 1.15 (there was a bug in tf2onnx 1.15 which required downgrade of onnx to 1.14)
- onnx 1.15.0 and tf2onnx 1.16.1

onnx runtime  
- 1.16.20231009.2.2a1fd25  
- 1.16.20231119.2.2ac381c  
- 1.18.20240126.1.fc44f96  

### Urgency

We need DirectML to be able to run inference on different hardware devices and drivers. So right now it seems a blocker as other alternatives require specific intel/nvidia/amd SDKs to be installed.

### Platform

Windows

### OS Version

Windows 10 22H2

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.18.20240126.1.fc44f96

### ONNX Runtime API

C

### Architecture

X64

### Execution Provider

Default CPU, CUDA, DirectML

### Execution Provider Library Version

DirectML 1.13.1.0

**Comment by yufenglee** [comment]
@fdwr, could you please take a look this issue?

**Comment by fdwr** [comment]
Attention to @smk2007 or @PatriceVignola.

**Comment by brookman1** [comment]
Why is the same result expected?  The computation is randomized: depending on the size of the dataset, deep learning models can come to wildly different fit with different initial random weights.

Random shuffle all keys, and load all but the last layer adds enough randomization.

Do you get close to the same results on the same hardware with different runs?  If you run from scratch end-to-end twice?

I'm trying the kaggle "Fastgan, the Pytorch FastGAN Notebook for TOMJERRY". This one starts off with some random 10% of the image dataset, never converging the same way twice, sometimes it does not converge, using the colab free tier T4 instance each time.
